bleak/_api.py

```diff
@@ -0,0 +1,237 @@
+import os
+import platform
+from bleak.exc import BleakError
+from bleak.backends.scanner import BaseBleakScanner
+from bleak.backends.client import BaseBleakClient
+
+
+_on_rtd = os.environ.get("READTHEDOCS") == "True"
+
+if _on_rtd:
+    from bleak.backends.scanner import (
+        BaseBleakScanner as _BleakScannerImplementation,
```

Comment:
To make the docs correct, all of the doc strings of the abstract base classes would need to be updated to document the concrete implementation rather than the abstract base class. Is this your intention?

Reply:
I was planning to burn that bridge when we get there, but apparently we have:-)

We really need two abstract base classes: one to specify how things look for the user and another one to specify how things look for a backend implementor. So the former wouldn't include the methods that the end user isn't supposed to call, for example. But I've experimented a bit with it and I keep getting into trouble with MRO...

Suggestions?

Reply:
Okay, I think I have something that works. I'm going to create another module `abstract_api.py` that has the user-visible abstract classes (and their user-visible documentation). the abstract backend classes are going to be subclassed from these (and include the methods and documentation for backend developers). When `_on_rtd` we inherit from the `abstract_api` classes.

But the backend documentation section will still include the "hidden" stuff.

Now only the constructors for `BleakClient` and `BleakScanner` need to be documented in a funny place.

Reply:
Implemented and pushed. Please check.

Reply:
> Suggestions?

I suppose a bit too late now :smile: but my first thought on this was to do something more like dependency injection instead of multiple inheritance. For example, in `BleakClient` might look something like this:

```python
from bleak.backends.client import get_platform_backend

...

class BleakClient:
    def __init__(self, ...):
        self._backend = get_platform_backend()
        ...

    async def connect(self):
        # common code like: if already connected, raise exception
        ...
        await self._backend.connect()
```

Reply:
> When _on_rtd we inherit from the abstract_api classes.

This works for the oneline docs, but what about `help()` in Python and docs for IDE intellesense? Wouldn't they still be using the doc string from one of the backends? So we would still need to write all of the docs strings 5 times?

Reply:
> my first thought on this was to do something more like dependency injection instead of multiple inheritance.

This would work only for `BleakClient` and `BleakScanner`, right? But there is definitely something to be said for it. 

It would also provide a nice hook for something that I need: using bleak in a framework that is not `asyncio`-based. I have something but it's hand-coded and not ready for prime time.

My suggestion is to keep this in mind as a separate issue. But it's your call...

Reply:
> what about help() in Python and docs for IDE intellesense?

I've completely ignored them:-)

Will check, and I think the writing the doctoring 5 times may only be needed for some very select methods, I think most can just use the doctoring of the abstract base class.

I'll check, but I will need help if things need fixing (don't know anything about the Android and Windows BLE stacks)

Reply:
> This would work only for BleakClient and BleakScanner, right?

Possibly. I haven't really looked at the other classes. But I think we should eventually be able to make most, if not all, of the other classes not backend-specific, so we could just ignore those for now.

We don't need to refactor the common code from the various backends just yet, but would it be too much work to just make the new `BleakClient` and `BleakScanner` classes a thin wrapper around each backend at this point in time (each method is implemented by `return self._backend.method()`) ?